test/functional/test_framework/test_framework.py

```diff
@@ -155,6 +159,28 @@ def main(self):
             exit_code = self.shutdown()
             sys.exit(exit_code)
 
+    def run_custom_tests(self):
+        incorrect_methods, uncallable_methods, correct_methods = [], [], []
+        for method_name in self.options.test_methods:
+            method = getattr(self, method_name, None)
+            if method is None:
+                incorrect_methods.append(method_name)
+            elif not callable(method):
+                uncallable_methods.append(method_name)
+            else:
+                correct_methods.append(method)
+
+        for method in correct_methods:
+            method()
+
+        if incorrect_methods or uncallable_methods:
+            error_message = f"Errors encountered:"
+            if incorrect_methods:
+                error_message += f"\nIncorrect methods (not found): {incorrect_methods}"
+            if uncallable_methods:
+                error_message += f"\nUncallable methods: {uncallable_methods}"
+            raise RuntimeError(error_message)
```

Comment:
The approach here was added to address [this comment](https://github.com/bitcoin/bitcoin/pull/30991#discussion_r1827274569) cc @rkrux .

Which is an overkill for this simple feature honestly.

Is the simplified approach below preferable?

```python3
def run_custom_tests(self):
    for method_name in self.options.test_methods:
        self.log.info(f"Attempting to execute method: {method_name}")
        method = getattr(self, method_name)
        method()
        self.log.info(f"Method '{method_name}' executed successfully.")
```

This version has a few more advantage. 
- It logs each method's execution attempt and relies on Python’s built-in exceptions to handle issues.
-  The code is simpler. 
- Additionally, it provides helpful feedback if a method name is close to a valid method but has a typo. For instance, omitting the final "n" in the method name would result in:
when running the example in the OP will result in

```
Traceback (most recent call last):
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/test_framework.py", line 165, in run_custom_tests
    method = getattr(self, method_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'ReindexTest' object has no attribute 'continue_reindex_after_shutdow'. Did you mean: 'continue_reindex_after_shutdown'?
```

Reply:
Since [this commit](https://github.com/bitcoin/bitcoin/commit/cc21876b125930f8320dbb95016f9ee7c1ffec55) was checking for the presence and call-ability of the passed methods already, the main intent of [this comment](https://github.com/bitcoin/bitcoin/pull/30991#discussion_r1827274569) was to add on to it by allowing the execution of all the methods in case of failure of any one of them.

> test-only dev-only debug-only feature

I believe keeping this^ in mind, it makes sense to just throw the error in case if any invalid method is passed. I agree the approach shared in the above comment is simpler and good enough for the use case.

Though, I feel sticking to only one term between `test_methods` or `custom_tests` would be better from readability POV.

Reply:
With the above approach, I see the following outputs that seem okay.

Error when a non existing attribute is passed.
```
2024-11-11T06:12:11.578000Z TestFramework (INFO): Attempting to execute method: test_args_l
2024-11-11T06:12:11.579000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
    ~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 165, in run_custom_tests
    method = getattr(self, method_name)
AttributeError: 'ConfArgsTest' object has no attribute 'test_args_l'. Did you mean: 'test_args_log'?
```

Error when a non callable attribute is passed.
```
2024-11-11T06:16:15.078000Z TestFramework (INFO): Attempting to execute method: chain
2024-11-11T06:16:15.078000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_custom_tests()
    ~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/rkrux/projects/bitcoin/test/functional/test_framework/test_framework.py", line 166, in run_custom_tests
    method()
    ~~~~~~^^
TypeError: 'str' object is not callable
```

Reply:
Done in latest push.
thanks